core/src/main/java/org/kohsuke/stapler/EvaluationTrace.java

```diff
@@ -61,4 +67,38 @@ public static EvaluationTrace get(StaplerRequest req) {
      * Used for counting trace header.
      */
     private static final String KEY = EvaluationTrace.class.getName();
+
+    public abstract static class ApplicationTracer {
+        protected abstract void record(StaplerRequest req, String message);
+
+        public static void trace(StaplerRequest req, String message) {
+            List<ApplicationTracer> tracers = getTracers();
+            for (ApplicationTracer tracer : tracers) {
+                tracer.record(req, message);
+            }
+        }
+
+        private static volatile List<ApplicationTracer> tracers;
+
+        @Nonnull
+        private static List<ApplicationTracer> getTracers() {
+            synchronized (ApplicationTracer.class) {
+                if (tracers == null) {
+                    List<ApplicationTracer> t = new ArrayList<>();
```

Comment:
You do not need a separate variable. You can initialize `tracers` directly, since it is protected by a synchronization barrier.